analyzers/src/SonarAnalyzer.CSharp/Rules/NamedPlaceholdersShouldBeUnique.cs

```diff
@@ -0,0 +1,122 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2024 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class NamedPlaceholdersShouldBeUnique : SonarDiagnosticAnalyzer
+{
+    private const string DiagnosticId = "S6677";
+    private const string MessageFormat = "Message template placeholder '{0}' is not unique.";
+
+    private static readonly HashSet<string> MicrosoftExtensionsLoggingMethods =
+        [
+            "Log",
+            "LogCritical",
+            "LogDebug",
+            "LogError",
+            "LogInformation",
+            "LogTrace",
+            "LogWarning"
+        ];
+
+    private static readonly DiagnosticDescriptor Rule = DescriptorFactory.Create(DiagnosticId, MessageFormat);
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+        ImmutableArray.Create(Rule);
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                if (c.Node is InvocationExpressionSyntax invocation
+                    && TemplateArgument(invocation, c.SemanticModel) is { } argument
+                    && MessageTemplates.Parse(argument.Expression.ToString()) is { Success: true } result)
+                {
+                    RaiseAtDuplicates(c, argument, result.Placeholders);
```

Comment:
This is probably the point where we can use composition.
We need a list of "something" that contains a rule diagnostic and a `ShouldRaise` method, and we loop over it.